src/query/plan/rewrite/index_lookup.hpp

```diff
@@ -136,21 +136,43 @@ class IndexLookupRewriter final : public HierarchicalLogicalOperatorVisitor {
       op.all_filters_ = std::move(leftover_filters);
     }
 
-    // edge uniqueness filter comes always before filter in plan generation
-    LogicalOperator *input = op.input().get();
-    LogicalOperator *parent = &op;
-    while (input->GetTypeInfo() == EdgeUniquenessFilter::kType) {
-      parent = input;
-      input = input->input().get();
-    }
-    bool is_child_cartesian = input->GetTypeInfo() == Cartesian::kType;
-
-    if (is_child_cartesian && removal.did_remove) {
-      // if we removed something from filter in front of a Cartesian, then we are doing a join from
-      // 2 different branches
-      auto *cartesian = dynamic_cast<Cartesian *>(input);
-      auto indexed_join = std::make_shared<IndexedJoin>(cartesian->left_op_, cartesian->right_op_);
-      parent->set_input(indexed_join);
+    // Filters are pushed down as far as they can go.
+    // If there is a Cartesian after, that means that the filter is working on data from both branches.
+    // In that case, we need to convert the Cartesian into a Join
+    if (removal.did_remove) {
+      LogicalOperator *input = op.input().get();
+      LogicalOperator *parent = &op;
+
+      // Find first possible branching point
+      while (input->HasSingleInput()) {
+        parent = input;
+        input = input->input().get();
+      }
+
+      std::vector<Symbol> modified_symbols;
+      for (const auto &filter : op.all_filters_) {
+        if (filter.property_filter) {
+          modified_symbols.push_back(filter.property_filter->symbol_);
+        }
+      }
```

Comment:
Building `modified_symbols` before we know we need it.
Prefer to be done only if `is_child_cartesian == true`
Atm `does_modify` is O(n^2), your call on if you change it the number of symbols should be small that it shouldn't ever matter.